### DISCUSSION TITLE
A standalone relocatable ImageMagick for current macOS

### DISCUSSION THREAD

**Discussion by emendelson** [ImageMagick/ImageMagick#7054] [General]
The question sometimes comes up - how to build a static-linked ImageMagick for macOS. It's probably impossible to build a static-linked executable, but the next best thing is portable self-contained macOS app that contains the executable and all the required dylibs in a single app bundle. You can simply run the executable inside the app from the command line, and it finds all the dylibs inside the same app. It's easy to create a symlink that makes it possible to run the executable without typing out the full path.

This can be used on systems that don't have Homebrew installed, and it doesn't need Homebrew to run.

Thanks to help given by people who are a lot more expert than I am, I was able to explain how to build one, and to provide a link to a .dmg that contains a notarized, universal app, plus a script that creates a symlink making it possible to run the portable `magick` from the command line. You can find it here. I hope it's useful to someone.

[https://mendelson.org/imagemagick.html](https://mendelson.org/imagemagick.html)

**Comment by fmw42** [comment]
Does the "magick" version at https://imagemagick.org/script/download.php do what you want?

**Reply by emendelson** [reply]
I'm confident that that version would work for anyone who goes through all the many steps required to set it up on their own system, but it doesn't make it possible to create a portable application that anyone can run on any system. The whole point of my instructions was to create a version that anyone can use without any setup at all.

**Reply by fmw42** [reply]
Sorry, I have not used it on my Mac.  Perhaps it is only for Linux and does not work on Max OSX platform.  A developer would need to comment further.

**Comment by fmw42** [comment]
Search "standalone" on the Development forum.  For example:  
https://github.com/ImageMagick/ImageMagick/discussions/5215

**Reply by emendelson** [reply]
Exactly! The only answer in that thread that actually works is my own build - the same one in the first post in this thread. I posted it originally on StackOverflow, and then created a separate page for it.

I searched every thread on the forum that purported to provide a standalone build - and some of those may have worked on ancient versions of macOS. But I believe the only working answer for recent versions of macOS is the build I pointed to it in the last post of the thread that you linked to in your reply.

**Comment by leleliu008** [comment]
@emendelson what does `A standalone ImageMagick` mean here? It seems that ImageMagick is not relocatable.

**Reply by emendelson** [reply]
@leleliu008 - Please read the first post in this thread, where everything is explained.  "Standalone" means that the macOS app bundle (ImageMagick.app) does not require any external dependencies. You can move it anywhere in your system.

**Reply by leleliu008** [reply]
That means A app should self-contained + relocatable.

**Reply by emendelson** [reply]
@leleliu008 - It is self-contained, and it is relocatable.

Have you actually tried it? You will see that it does exactly what I say. You can locate it anywhere and run the enlosed executable from the command line.

Please do not explain why it does not work until you have actually tried it and found that it does work.

**Reply by leleliu008** [reply]
> @leleliu008 - Please read the first post in this thread, where everything is explained.

Your first post in this thread just talks about self-contained, dosn't motion relocatable.

**Comment by leleliu008** [comment]
https://github.com/leleliu008/xcpkg may help folks who want to build a self-contained ImageMagick.

**Reply by emendelson** [reply]
@leleliu008 - Have you actually tried using it to build ImageMagick?

**Reply by leleliu008** [reply]
@emendelson I build it just now. https://github.com/leleliu008/uppm-package-repository-macos-14.0-arm64/releases/download/2024.05.27/imagemagick-7.1.1.32-macos-14.0-arm64.tar.xz

**Reply by emendelson** [reply]
@leleliu008 - Your build starts in the command line, but it won't convert a .PNG file to another format. It gives this error messaga:

`magick: NoDecodeDelegateForThisImageFormat `PNG' @ error/constitute.c/ReadImage/746.'

The build that I posted in the linked page on my site converts successfully.

**Comment by emendelson** [comment]


**Reply by leleliu008** [reply]
Yes, I don't known why. That's why I say It seems that it is not relocatable.

**Reply by emendelson** [reply]
@leleliu008 - Your build is not relocatable. My app is relocatable. Have you actually tried the app that I link to on my web page? Please do not say that it does not work until you have tried it.

**Reply by emendelson** [reply]
@leleliu008 - And here is the link if you do not want to read down to the bottom of the page:

https://mendelson.org/PortableImageMagickInstaller.dmg

Again, please do not say it does not work until you have tried it and found that it works exactly as it should.

**Reply by leleliu008** [reply]
@emendelson I didn't try yours. I say this software. I do not refer to your distrobution.

**Reply by emendelson** [reply]
@leleliu008 - What you say about the original ImageMagick is course correct. That is why I created my build and app.

**Reply by leleliu008** [reply]
Yes, I refer to the original ImageMagick. I tried the original ImageMagick.

**Reply by leleliu008** [reply]
I do not find your build script. as per https://imagemagick.org/script/download.php , It should be relocatable.

**Reply by emendelson** [reply]
@leleliu008 Please look at the page I link to at the top of this thread:

https://mendelson.org/imagemagick.html

You can create a build script from the steps listed there. I did not create a single script because different architectures and different macOS versions require different commands. If you can create a script that would choose the correct options for different versions and different architectures, that would be very helpful.

**Reply by leleliu008** [reply]
@emendelson Thanks. I see it. I do not find any paticular steps to make it relocatable. I do not use HomeBrew. I build ImageMagick from source, I just test it whether is relocatable, I encountered the problem, I started to search on google and I found this post.

**Comment by emendelson** [comment]
@leleliu008 - The app is relocatable becuase all the dylibs are included in the app bundle and `dylibbundler` tells the executable that it can find the dylibs in the bundle. The executable itself cannot be relocated outside the bundle.

**Reply by leleliu008** [reply]
my build is also self-contained.

**Reply by emendelson** [reply]
@leleliu008 - Yes, but it doesn't have the required delegate for PNG files, and perhaps for other formats also, so many users won't be able to use it at all. I don't think either of us needs to say anything more about this subject.

**Comment by leleliu008** [comment]
After reading this https://imagemagick.org/script/resources.php#environment , it works by setting environment variables:

```bash
curl -LO https://github.com/leleliu008/uppm-package-repository-macos-14.0-arm64/releases/download/2024.05.27/imagemagick-7.1.1.32-macos-14.0-arm64.tar.xz
tar xf imagemagick-7.1.1.32-macos-14.0-arm64.tar.xz
cd imagemagick-7.1.1.32-macos-14.0-arm64

export MAGICK_HOME="$PWD"
export MAGICK_CODER_MODULE_PATH="$PWD/lib/ImageMagick-7.1.1/modules-Q16HDRI/coders"
export MAGICK_CONFIGURE_PATH="$PWD/etc/ImageMagick-7"

bin/magick 1.jpg 2.png
```

**Reply by emendelson** [reply]
@leleliu008 - Excellent work. Of course, this will not help if you copy your executable to another machine. My app is designed to work anywhere, with no variables set.  I will be very interested to see an executable that is as portable as my app is.

**Reply by leleliu008** [reply]
@emendelson Your distribtion is also excellent. I do not like .dmg file. I like tarball file. because I usually use it in shell script. My is good enough to use for me. I do not want to create a universal App.

**Comment by nagualcode** [comment]
@emendelson You Sir are a Gentleman and a Scholar

**Comment by ylazy** [comment]
Today when I downloaded ImageMagick for MacOS I thought it's a standalone bundle until I got errors.
I don't want to install ImageMagick, I need to include it in my app so the package that I downloaded is not usable.
I searched a lot and found this thread.
Thank you for the .dmg file. I can extract it to find what I need.

**Comment by ylazy** [comment]
Hi @emendelson ,

With your latest magick update, is it safe to remove the symlink and move the Magick.app to another location, for example including it in my own app? Is there any limitation or something else?

After moving it to another location:

<img width="1469" height="406" alt="image" src="https://github.com/user-attachments/assets/58cb921b-f18a-46df-931b-df98f930dbec" />

I saw the ? icon. But it seems to work properly.

Thanks!

**Comment by emendelson** [comment]
@ylazy - I haven't tried this, so I don't know. If you don't find any problem, then it clearly works!
